### PR TITLE
TST: isinstance(OptimizeResult.message, str) closes #13001

### LINKED ISSUE #13001: `OptimizeResult.message` from `L-BFGS-B` is a bytes, not a str.
The doc for `OptimizeResult` says that its `message` attribute will be a string.

For `L-BFGS-B`, it is a bytes object.   That makes it harder to serialize the `OptimizeResult`.

#### Reproducing code example:

```
from scipy.optimize import minimize, rosen
x0 = [1.3, 0.7, 0.8, 1.9, 1.2]
for method in ('Nelder-Mead', 'Powell',  'BFGS',
               'TNC', 'COBYLA', 'SLSQP', 'L-BFGS-B'):
    res = minimize(rosen, x0, method=method, tol=1e-6)
    print(method, type(res.message), res.message)
```

prints out

```
Nelder-Mead <class 'str'> Optimization terminated successfully.
Powell <class 'str'> Optimization terminated successfully.
BFGS <class 'str'> Desired error not necessarily achieved due to precision loss.
TNC <class 'str'> Converged (|f_n-f_(n-1)| ~= 0)
COBYLA <class 'str'> Maximum number of function evaluations has been exceeded.
SLSQP <class 'str'> Optimization terminated successfully
L-BFGS-B <class 'bytes'> b'CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH'
```

#### Error message:

N/A

#### Scipy/Numpy/Python version information:
```
>>>import sys, scipy, numpy; print(scipy.__version__, numpy.__version__, sys.version_info)
1.5.2 1.19.1 sys.version_info(major=3, minor=8, micro=5, releaselevel='final', serial=0)
```

### PR DESCRIPTION
**Pull request by andyfaff**
Tests that all `isinstance(OptimizeResult.message, str)`. Fixes minimizer methods that don't comply.

Closes #13001

### CODE REVIEW

**Comment by andyfaff** [comment]
closes #13001

**Comment by ilayn** [comment]
+1 though the messages still reads a bit cryptic. 

By the way, moved the Github closing instruction to the description to make it work.

**Comment by andyfaff** [comment]
The LBFGS message is cryptic, or the commit message is cryptic?

**Comment by ilayn** [comment]
The LBFGS messages but I guess they have been there for quite a while now.

File: scipy/optimize/tests/test_optimize.py [hunk 1]
```diff
@@ -1891,9 +1891,12 @@ def test_attributes_present(self):
                 if method in skip and attribute in skip[method]:
                     continue
 
-                assert_(hasattr(res, attribute))
+                assert hasattr(res, attribute)
```
Comment (mdhaber) [inline scipy/optimize/tests/test_optimize.py:1894]: Just curious - why the change?

**Review by mdhaber — APPROVED** [review]
Looks good except for the change to `assert` statements; I thought the function was preferred in unit tests.

**Comment by andyfaff** [comment]
plain assert is ok now, it doesn't get optimised away. I prefer this style.

**Comment by newville** [comment]
@andyfaff Thanks for the quick response!

File: scipy/optimize/tests/test_optimize.py (hunk 1 above)
Comment (tylerjereddy) [reply scipy/optimize/tests/test_optimize.py:1894]: I think the "informal decision" the last time this came up was that it was "ok," perhaps encouraged, to change from the old-style function to the plain `assert` now that we use `pytest` (which has shims to prevent optimizing out of asserts) in cases where you happen to see it or add a new usage, but that we probably don't want to do a mass auto-reformat.
